File: client/src/services/coupon.service.ts
import { apiService } from './api';

export interface Coupon {
  couponId: number;
  code: string;
  description: string;
  discountPercent: number;
  requiredPoints: number;
  startDate: string;
  endDate: string;
  isActive: boolean;
  canUse?: boolean; // Whether user has enough points to use this coupon
}

export interface CouponValidationResponse {
  valid: boolean;
  coupon?: Coupon;
  message?: string;
  discountAmount?: number;
}

export const couponService = {
  // Get all available coupons (with loyalty points check if authenticated)
  getAvailableCoupons: async (): Promise<Coupon[]> => {
    try {
      const coupons = await apiService.get<Coupon[]>('/coupons/available');
      // Sort: coupons user can use (canUse = true) first, then by discountPercent desc
      return coupons.sort((a, b) => {
        if (a.canUse && !b.canUse) return -1;
        if (!a.canUse && b.canUse) return 1;
        return b.discountPercent - a.discountPercent;
      });
    } catch (error) {
      console.error('Error fetching available coupons:', error);
      return [];
    }
  },
  
  // Validate coupon with loyalty points
  validateCouponWithPoints: async (couponId: number, totalAmount: number): Promise<CouponValidationResponse> => {
    try {
      return await apiService.post<CouponValidationResponse>(
        `/coupons/${couponId}/validate?totalAmount=${totalAmount}`,
        {}
      );
    } catch (error: any) {
      console.error('Error validating coupon:', error);
      return {
        valid: false,
        message: error?.response?.data?.message || 'Không thể validate mã giảm giá',
      };
    }
  },

  // Validate coupon code (client-side validation)
  validateCoupon: async (code: string, totalAmount: number): Promise<CouponValidationResponse> => {
    try {
      // Get coupon information from backend
      const coupon = await apiService.get<Coupon>(`/coupons/code/${code}`);
      
      if (!coupon) {
        return {
          valid: false,
          message: 'Mã khuyến mãi không tồn tại',
        };
      }

      // Validate on client-side
      const now = new Date();
      const startDate = new Date(coupon.startDate);
      const endDate = new Date(coupon.endDate);

      // Check if coupon is active
      if (!coupon.isActive) {
        return {
          valid: false,
          message: 'Mã khuyến mãi không còn hiệu lực',
        };
      }

      // Check date validity
      if (now < startDate) {
        return {
          valid: false,
          message: 'Mã khuyến mãi chưa có hiệu lực',
        };
      }

      if (now > endDate) {
        return {
          valid: false,
          message: 'Mã khuyến mãi đã hết hạn',
        };
      }

      // Không cần check minimum order value nữa - coupon áp dụng cho tất cả đơn hàng

      // Calculate discount
      const discountAmount = (totalAmount * coupon.discountPercent) / 100;

      return {
        valid: true,
        coupon,
        message: 'Mã khuyến mãi hợp lệ',
        discountAmount,
      };
    } catch (error: any) {
      console.error('Error validating coupon:', error);
      return {
        valid: false,
        message: 'Mã khuyến mãi không tồn tại',
      };
    }
  },

  // Get coupon by code
  getCouponByCode: async (code: string): Promise<Coupon | null> => {
    try {
      return await apiService.get<Coupon>(`/coupons/code/${code}`);
    } catch (error) {
      console.error('Error fetching coupon by code:', error);
      return null;
    }
  },
};

File: server/src/main/java/iuh/fit/server/services/CouponService.java
package iuh.fit.server.services;

import iuh.fit.server.dto.request.CouponRequest;
import iuh.fit.server.dto.request.CouponValidationRequest;
import iuh.fit.server.dto.response.CouponResponse;
import iuh.fit.server.dto.response.CouponValidationResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface CouponService {

    /**
     * Validate coupon code and check if it can be applied to the order
     * 
     * @param request Validation request containing code and total amount
     * @return Validation response with coupon details if valid
     */
    CouponValidationResponse validateCoupon(CouponValidationRequest request);

    /**
     * Get all active and valid coupons
     * 
     * @return List of active coupons
     */
    List<CouponResponse> getAvailableCoupons();

    /**
     * Get coupon by code
     * 
     * @param code Coupon code
     * @return Coupon response if found
     */
    CouponResponse getCouponByCode(String code);

    /**
     * Get available coupons with user's loyalty points check
     * 
     * @param userId User ID (optional, null for guest)
     * @return List of coupons with canUse flag
     */
    List<CouponResponse> getAvailableCoupons(Integer userId);

    /**
     * Validate coupon with user's loyalty points
     * 
     * @param couponId    Coupon ID
     * @param userId      User ID
     * @param totalAmount Order total amount
     * @return Validation response
     */
    CouponValidationResponse validateCouponWithPoints(Integer couponId, Integer userId, Double totalAmount);

    /**
     * Get all coupons with pagination
     * 
     * @param pageable Pagination information
     * @return Page of coupons
     */
    Page<CouponResponse> getAllCoupons(Pageable pageable);

    /**
     * Search coupons by code or description
     * 
     * @param query    Search query
     * @param pageable Pagination information
     * @return Page of matching coupons
     */
    Page<CouponResponse> searchCoupons(String query, Pageable pageable);

    /**
     * Get coupon by ID
     * 
     * @param id Coupon ID
     * @return Coupon response
     */
    CouponResponse getCouponById(Integer id);

    /**
     * Create new coupon
     * 
     * @param request Coupon request
     * @return Created coupon
     */
    CouponResponse createCoupon(CouponRequest request);

    /**
     * Update existing coupon
     * 
     * @param id      Coupon ID
     * @param request Coupon request
     * @return Updated coupon
     */
    CouponResponse updateCoupon(Integer id, CouponRequest request);

    /**
     * Delete coupon
     * 
     * @param id Coupon ID
     */
    void deleteCoupon(Integer id);
}
